### ISSUE TITLE
(General) Advantages using Python for stats and plotting?

### ISSUE THREAD

**Issue by mwychan-BME** [BME1478H/Fall2020class#264]
Hi, just a general question (feel free to ignore if busy). 

What are the pros and cons of using Python for stats (from t-test/ANOVA to PCA/KMS) and plotting data? As compared to using a different language, like R? Are there certain use cases where Python has significant strengths? I understand it might boil down to being a preference thing, but personal opinions are appreciated!

**Comment by RayNele** [comment]
Hey @mwychan-BME thanks for the question, PERSONALLY I use python in all cases just because I'm most comfortable with it (and everything is well documented and easy to understand). I don't use R or Matlab anywhere near as often, but here is a breakdown of my best understanding of the 3 big data science languages. Keep in mind these are my personal opinions, and by no means do I mean to turn you away from any of these languages. 

Python is a pretty simple language to learn and many of the most common stats/graphing/data science solutions are available as packages like numpy, pandas, matplotlib, etc. I would use python hands down with large sets of data (I'm talking in the tens-hundred thousands) or if the data contains strings (python handles those very well, I did my undergrad in psych, so there's a lot of qualitative as well as quantitative stuff that needs to be processed here). Also Python (R as well actually) is open source, and anyone is free to contribute. There are tons and tons of (open source) libraries and packages out there, some created by other python users just for fun. For example, [here's a color scheme](https://github.com/dhaitz/mplcyberpunk for matplotlib someone made just the other day. If you wanted to do some task, it's likely someone has written a package for it on python. 

As for R, R is very much a statistician's language. Anything that math people have found interesting to study or simulate, it has probably been implemented in R. Also I'd say (personal opinion) that R is slightly harder to pick up compared to python in that it has slightly confusing syntax. If I needed to do work on "insert obscure math pattern or stats test here" I would use an R library instead of writing it up from scratch on Python. 

Last but not least, Matlab. First of all matlab is commercial, so to use it, you'd need a license (which I think regular people like us can't afford to just play around in it without the backing of an institution). I'd say for a fresh programmer, matlab is the easiest to pick up out of all 3. I've often heard matlab called "the programming language for people who don't know how to code". There are a ton of weird quirks with it that you wouldn't see in other languages. For example, matlab starts counting from 1. Who starts counting from 1??? Everyone (programmers) knows the first number is 0. HOWEVER, it is useful in that you can model/compute extremely complicated things with little to no experience with matlab or coding in general. 

Keep in mind that these advantages/disadvantages go both ways. Python and its packages are open source, so in a way, they are tested and upheld by the community (That's us!). However, matlab again, is commercial, so employees are paid to do all the testing work. There's no clear winner here, Matlab has better support since it's all created by the same group of people; library X isn't guaranteed to work with library Y in python (although numpy, pandas, and matplotlib, all play very nice with each other). 

As for computational speed (just an overarching generalization) I believe it goes MATlab > Python > R, but again, this really doesn't matter for users like us. 

BASICALLY to sum it all up: It's all the same for casual users like us unless you're very deep into it and doing development work for X company. I'm sure if you asked some of the other TA's or other CS students you'll find they have differing opinions but for the scope of what we're likely to be doing, I don't think Python has any limitations.
